### ISSUE TITLE
org-wild-notifier-check breaks navigation in org agenda

### ISSUE THREAD

**Issue by codygman** [akhramov/org-wild-notifier.el#10]
Emacs 26
Org 9

1. Open org agenda
2. Navigate to any items successfully with tab or enter
3. Org-wild-notifier-check
4. Navigation to a items in org agenda broken until you refresh it or bury/restart it

**Comment by codygman** [comment]
I installed from melpa which might not be up to date since i think this commit fixes the issue:

https://github.com/akhramov/org-wild-notifier.el/commit/6453ea6556f7425ce79c748dc4962f73421e09df

**Comment by akhramov** [comment]
Hi, thanks for the report!

What version do you have installed? 6453ea6 is included in the latest version (0.2.4)

**Comment by codygman** [comment]
@akhramov 

```
;; Version: 0.2.4
;; Package-Version: 20180222.425
```

I'm wondering if we need to use org agenda sticky views:

```
By default, Org maintains only a single agenda buffer and rebuilds it each time you change the view, to make sure everything is always up to date. If you switch between views often and the build time bothers you, you can turn on sticky agenda buffers (make this the default by customizing the variable  org-agenda-sticky). With sticky agendas, the dispatcher only switches to the selected view, you need to update it by hand with r or g. You can toggle sticky agenda view any time with org-toggle-sticky-agenda.
```

I feel like this error happens because the `org-agenda-list` call's agenda somehow messes with any existing org-agendas. It sounds like sticky agendas might prevent that from happening?

Later I'll see if I can write a failing test to make this easier to reproduce/verify it's reproducible.

**Comment by codygman** [comment]
Or perhaps just naming the new agenda org-wild-notifer opens something unique.

**Comment by codygman** [comment]
I also found out that running `org-wild-notifier-check` in an org mode buffer where you've used `org-narrow-to-subtree` causes the org mode file to widen.

**Comment by codygman** [comment]
I've verified that using sticky views fixes this issue:

```org
*** TODO see if org agenda sticky views fixes issue
    :LOGBOOK:
    CLOCK: [2018-08-12 Sun 21:57]
    :END:
**** reproduce issue
***** disable sticky views
***** prove its disabled
      #+BEGIN_EXAMPLE
        org-agenda-sticky is a variable defined in ‘org-agenda.el’.
        Its value is nil
      #+END_EXAMPLE
***** =SPC a o a=
***** =org-agenda-switch-to= worked
***** call org-wild-notifier-check
***** gave "trying to switch to non-existent buffer
**** try solution
***** enabled sticky views
***** prove its enabled
      #+BEGIN_EXAMPLE
        org-agenda-sticky is a variable defined in ‘org-agenda.el’.
        Its value is t
        Original value was nil
      #+END_EXAMPLE
***** navigated to something
***** called org-wild-notifier-check
***** =org-agenda-switch-to= worked
**** it works, but will enabling sticky buffers for this function change normal functionality of org-agenda?
**** If I try to run an org agenda command will it open in a new buffer if sticky views is activated because org-wild-notifier-check ran?
```

I wasn't able to make a test though because I wasn't sure how to set my agenda files to the test file and run the tests.

**Comment by codygman** [comment]
I made this change:

```
(defun org-wild-notifier-check ()
  "Parse agenda view and notify about upcomming events."
  (interactive)
  (save-window-excursion
    (let ((org-agenda-use-time-grid nil)
          (org-agenda-compact-blocks t)
          (org-agenda-sticky t)
          (org-agenda-window-setup 'current-window)
          (org-agenda-buffer-name nil)
          (org-agenda-buffer-tmp-name org-wild-notifier--agenda-buffer-name))

      (org-agenda-list 2)

      (-each
        (->> (org-wild-notifier--retrieve-events)
             (-map 'org-wild-notifier--check-event)
             (-flatten)
             (-uniq))
        'org-wild-notifier--notify)

      (kill-buffer))))
```

But that wasn't enough. It looks like the only thing that fixes this issue for me is:

```
(setq org-agenda-sticky t)
```

and putting that in my init file.

**Comment by codygman** [comment]
It appears that even with `org-agenda-sticky` set to true each call of org-wild-notifier-check will negate `org-narrow-to-subtree` and I'm guessing `org-tree-to-indirect-buffer`.

If you can offer any direction or thoughts @akhramov I'll be looking more into this in the afternoon.

**Comment by akhramov** [comment]
Hi, @codygman, I'll take a look at this problem on the weekend. 

So far, I narrowed the problem to following:

##### Before `org-wild-notifier-check`:

`M-: (org-get-at-bol 'org-marker)` returns valid marker

##### After `org-wild-notifier-check`:

`M-: (org-get-at-bol 'org-marker)` returns invalid marker

**Comment by akhramov** [comment]
Huh... it looks like `org-wild-notifier-check` changes `org-agenda-buffer-name`. We can probably use `unwind-protect` to restore the original variable's value.

**Comment by codygman** [comment]
@akhramov Thanks for fixing this! I'll try to take a look tomorrow.

**Comment by codygman** [comment]
I'll try doing another test later, but I'm fairly sure I tested correctly and got:

```
Debugger entered--Lisp error: (wrong-type-argument integer-or-marker-p nil)
  org-agenda-goto()
  funcall-interactively(org-agenda-goto)
  call-interactively(org-agenda-goto nil nil)
  command-execute(org-agenda-goto)
```

after running `org-wild-notifer-check`.

**Comment by lgmoneda** [comment]
Does this #11 also aim to fix the `org-narrow-to-subtree` problem? I've tried it and it messes the narrowing yet.

**Comment by lgmoneda** [comment]
@akhramov it's the one described by @codygman:

> I also found out that running org-wild-notifier-check in an org mode buffer where you've used org-narrow-to-subtree causes the org mode file to widen.

So what happens is that when I use `narrow-to-subtree` the org-wild-notifier will trigger the check after a while and exit the narrow view automatically.

**Comment by akhramov** [comment]
Thanks, @lgmoneda. Seems like the solution proposed by me is not viable.

I'm going to invest some time into this in future, unless someone takes over sooner. Sorry for the inconvenience.

**Comment by codygman** [comment]
I think this Edna package has to do something similar to org wild notifiertifier to implement triggers. Maybe tge answer is in there:

http://www.nongnu.org/org-edna-el/#operation

**Comment by lgmoneda** [comment]
A workaround: disable and enable `org-wild-notifier` when using narrow and widen:

```
(defadvice org-narrow-to-subtree (before block-wild activate)
  (when (fboundp 'org-wild-notifier-mode)
    (org-wild-notifier-mode -1)))

(defadvice widen (before block-wild activate)
  (when (fboundp 'org-wild-notifier-mode)
    (org-wild-notifier-mode 1)))
```

**Comment by akhramov** [comment]
Alright, the patch issued by @neeasade (#14) seems to resolve this.
Feel free to comment here if you have any further issues.

**Comment by akhramov** [comment]
@codygman I'm sorry for closing this one

#17 resolves this issue (and the whole class of  similar problems), but I need to find a way to put the tests back on track. Meanwhile you can install package from the file... If you encounter any issues, I'll be happy to assist.

**Comment by lgmoneda** [comment]
The problem persists here even after updating org-wild-notifier. Am I missing something?

**Comment by akhramov** [comment]
Fixed in v0.3.0, please wait for MELPA archives to update.
Thanks for your patience :)
